matscipy/sliding_p.py

```diff
@@ -0,0 +1,299 @@
+
+Some parts are based on L. Pastewka, S. Moser, and M. Moseler, Tribol. Lett. 39, 49 (2010)
+as indicated again below.
+
+Author: Alexander Held
+"""
+
+from __future__ import (
+    division,
+    absolute_import,
+    print_function,
+    unicode_literals
+)
+import numpy as np
+from ase.units import kB, fs, GPa
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class AutoDamping(object):
+    """Automatic damping
+
+    Following L. Pastewka, S. Moser, and M. Moseler,
+    Tribol. Lett. 39, 49 (2010).
+    """
+
+    def __init__(self, C11, p_c=0.01):
+        """Constructor.
+
+        Arguments as described in L. Pastewka, S. Moser, and M. Moseler,
+        Tribol. Lett. 39, 49 (2010).
+        """
+        self.C11 = float(C11)
+        self.p_c = float(p_c)
+
+    def get_M_gamma(self, slider, atoms):
+        A = slider.get_A(atoms)
+        l = atoms.cell[slider.vdir, slider.vdir]
+        t_c = l / slider.v
+        omega_c = 2. * np.pi / t_c
+        h1 = atoms.positions[slider.top_mask, slider.Pdir].min()
+        h2 = atoms.positions[slider.bottom_mask, slider.Pdir].max()
+        h = h1 - h2
+        k = self.C11 * A / h
+        M = k * omega_c ** -2 * np.sqrt(self.p_c ** -2 - 1.)
+        # gamma = np.sqrt(2. * M * k)  # <- expression from paper, but this is wrong
+        gamma = 2. * np.sqrt(M * k)
+        return M, gamma
+
+
+class FixedDamping(object):
+    """Damping with fixed damping constant and fixed mass."""
+
+    def __init__(self, gamma, M_factor=1.0):
+        self.gamma = float(gamma)
+        self.M_factor = float(M_factor)
+
+    def get_M_gamma(self, slider, atoms):
+        M_top = atoms.get_masses()[slider.top_mask].sum()
+        return M_top * self.M_factor, self.gamma
+
+
+class FixedMassCriticalDamping(object):
+    """Damping with fixed mass and critical damping constant.
+
+    Useful for fast pressure equilibration with small lid mass.
+    """
+
+    def __init__(self, C11, M_factor=1.0):
+        self.C11 = float(C11)
+        self.M_factor = float(M_factor)
+
+    def get_M_gamma(self, slider, atoms):
+        M_top = atoms.get_masses()[slider.top_mask].sum()
+        M = M_top * self.M_factor
+        A = slider.get_A(atoms)
+        h1 = atoms.positions[slider.top_mask, slider.Pdir].min()
+        h2 = atoms.positions[slider.bottom_mask, slider.Pdir].max()
+        h = h1 - h2
+        k = self.C11 * A / h
+        gamma = 2. * np.sqrt(M * k)
+        return M, gamma
+
+
+class SlideWithNormalPressureCuboidCell(object):
+    """ASE constraint used for sliding with pressure coupling
+
+    Only works with diagonal cuboid cells so far.
+    Sliding only works along cell vectors so far.
+    Following L. Pastewka, S. Moser, and M. Moseler,
+    Tribol. Lett. 39, 49 (2010).
+    """
+
+    def __init__(self, top_mask, bottom_mask, Pdir, P, vdir, v, damping):
+        """Constructor.
+
+        top_mask -- boolean numpy array a with a[i] == True for i the index
```

Comment:
docstrings should be numpydoc formatted. See here: https://numpydoc.readthedocs.io/en/latest/format.html